### ISSUE TITLE
Conv2d 40x slowdown depending on its filter weight (on CPU)

### ISSUE THREAD

**Issue by AlexanderKazakov** [pytorch/pytorch#21734]
## 🐛 Bug

nn.Conv2d performance (on CPU) changes 40 times by only changing its filter weights

## To Reproduce

1. Download script, .pth files
2. Run script

## Results on my PC
Script output:
```
slow time: 8.078160285949707
slow time: 8.04417896270752
slow time: 8.085155010223389
fast time: 0.14791274070739746
fast time: 0.14391350746154785
fast time: 0.15190911293029785
slow time: 8.12812876701355
slow time: 8.035325050354004
slow time: 8.056166410446167
fast time: 0.15190744400024414
fast time: 0.14391160011291504
fast time: 0.14591288566589355
slow time: 8.197075843811035
slow time: 8.11712646484375
slow time: 8.094138145446777
fast time: 0.1449122428894043
fast time: 0.14391422271728516
fast time: 0.14391303062438965
```
Results depend on the Conv2d state only.
torch.no_grad() switching or changing input values do not affect the time.

## Expected behavior
Stable forward time

## Environment

- PyTorch version: 1.1.0
- Is debug build: No
- CUDA used to build PyTorch: Could not collect  # pytoch-cpu from anaconda
- OS: Windows 10
- GCC version: ERROR
- CMake version: ERROR
- Python version: 3.6
- Is CUDA available: No
- CUDA runtime version: ERROR
- GPU models and configuration: ERROR
- Nvidia driver version: ERROR
- cuDNN version: ERROR
-Versions of relevant libraries:
- [pip] 19.0.3
- [conda] 4.6.14
- How you installed PyTorch conda
- CUDA/cuDNN version NO GPU

## Additional context

I checked CPU profile while running the script, to be attached.
Also, I attach histograms of the layer's weight.flatten() histograms. There is a difference, but I don't see how it can affect performance so dramatically

All the files are here (because the archive is too large): https://drive.google.com/file/d/13H8viip0aXWhofNuHh9nX7_ezAbyc4Zv/view?usp=sharing

**Comment by AlexanderKazakov** [comment]
By the way, I place the script here in order to show that there is nothing special in it:
```
import torch
import os
from torch import nn
import time


def perf_test(postfix):
    inp = torch.load(os.path.join(base_dir, 'input_{}.pth'.format(postfix)))
    out = torch.load(os.path.join(base_dir, 'output_{}.pth'.format(postfix)))
    conv = nn.Conv2d(in_channels=64, out_channels=96, kernel_size=3, padding=1)
    conv.load_state_dict(torch.load(os.path.join(base_dir, 'conv2d_{}.pth'.format(postfix))))
    start = time.time()
    res = conv(inp)
    end = time.time()
    assert torch.all(res == out).item() == 1
    print(postfix, 'time:', end - start)


base_dir = 'results/conv2d_performance'
with torch.no_grad():  # does not matter either way
    for cntr1 in range(3):
        for cntr2 in range(3):
            perf_test('slow')
        for cntr2 in range(3):
            perf_test('fast')
```

Also, the CPU profile running the script
![proc](https://user-images.githubusercontent.com/5320488/59443463-b2a9da00-8e04-11e9-9228-56fa7378bd89.png)

And the histogram of the conv.weight.flatten():
*Fast:*
![weight_fast](https://user-images.githubusercontent.com/5320488/59443535-d2d99900-8e04-11e9-8585-6a513946578b.png)

*Slow:*
![weight_slow](https://user-images.githubusercontent.com/5320488/59443559-d9681080-8e04-11e9-8134-585b8e1e2b40.png)

**Comment by vadimkantorov** [comment]
denormalized floats to blame?

**Comment by ezyang** [comment]
I agree with @vadimkantorov, if the slow down is data dependent, then this is almost certainly due to denormalized floats. I don't know if PyTorch library has an easy way to check for this in your dataset, but that would be the first thing to check.

**Comment by AlexanderKazakov** [comment]
@vadimkantorov Thank you, you are right! When I checked the filter weight for denormalized values, in the fast filter there were not such at all, and in the slow one there were about 60% denormalized values. When I replaced them like
``` python
def clear_from_denormalized_values(model):
    eps = pow(2, -125)
    for p in model.parameters():
        p.detach()[torch.abs(p) < eps] = 0

...
clear_from_denormalized_values(conv)
...
```
forward times became equal and model behavior is the same.

Intersting, but not so important question is why eps = pow(2, -125)? From documentation it should be eps = pow(2, -126), but with the latter I had 6x slowdown anyway.
